validphys2/src/validphys/pineparser.py

```diff
@@ -292,6 +294,13 @@ def pineappl_reader(fkspec):
             if apfelcomb.get("shifts") is not None:
                 ndata += apfelcomb["shifts"][i]
 
+        # Here I need to provide the additional dimension to match the other FKs
+        # First I need to find where holes are
+        missing_x_points = np.setdiff1d(xgrid, p.x_grid())
+        missing_x_points_index = [list(xgrid).index(miss) for miss in missing_x_points]
+        for miss_index in missing_x_points_index:
+            raw_fktable = np.insert(raw_fktable, miss_index, 0., axis=2)
+            raw_fktable = np.insert(raw_fktable, miss_index, 0., axis=3)
```

Comment:
```suggestion
        missing_x_points = np.setdiff1d(xgrid, p.x_grid(), assume_unique=True)
        for x_point in missing_x_points:
            miss_index = list(xgrid).index(x_point)
            raw_fktable = np.insert(raw_fktable, miss_index, 0., axis=2)
            if hadronic:
                raw_fktable = np.insert(raw_fktable, miss_index, 0., axis=3)
```

A few considerations here, which I'm writing as a suggestion to order my own thoughts but please read through the changes to make sure I didn't break them.

- `assume_unique` ensures that the output is ordered, pineappl fktables are always ordered and the xgrids should be unique so this ensures that the result is ordered which is needed for:
- inserting things into an array makes me nervous because the results get shifted. `p.x_grid` is guaranteed to be a subset of `xgrid` so I think the only possibility for trouble is if the missing points are not ordered. So by looping over a list of the ordered points in `xgrid` that are not in `p.x_grid` you are guaranteed that they get inserted in the right order.
- since boths loops are essentially the same one I've joined them
- Finally, I'm guessing that axis=3 would only be active for double-hadronic observables.

Reply:
So I agree with points 1, 2 and 3. For point 4 I am not sure, let me check again.

Reply:
Yes you are right. I believe I can commit your suggestion safely. Thanks!